ff1.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+import sys; assert sys.version_info[0] >= 3, "Python 3 required."
+
+import os
+from binascii import unhexlify, hexlify
+
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from cryptography.hazmat.backends import default_backend
+
+from sapling_utils import bebs2ip, i2bebsp, beos2ip, bebs2osp, cldiv
+
+# Morris Dworkin
+# NIST Special Publication 800-38G
+# Recommendation for Block Cipher Modes of Operation: Methods for Format-Preserving Encryption
+# <http://dx.doi.org/10.6028/NIST.SP.800-38G>
+# specialized to the parameters below and a single-block PRF; unoptimized
+
+radix = 2
+minlen = maxlen = 88
+maxTlen = 255
+assert 2 <= radix and radix < 256
+assert radix**minlen >= 100
+assert 2 <= minlen and minlen <= maxlen and maxlen < 256
+
+NUM_2 = bebs2ip
+STR_2 = i2bebsp
+
+
+def ff1_aes256_encrypt(key, tweak, x):
+    n = len(x)
+    t = len(tweak)
+    assert minlen <= n and n <= maxlen
+    assert t <= maxTlen
+
+    u = n//2; v = n-u
+    assert u == v
+    A = x[:u]; B = x[u:]
+    assert radix == 2
+    b = cldiv(v, 8)
+    d = 4*cldiv(b, 4) + 4
+    assert d <= 16
+    P = bytes([1, 2, 1, 0, 0, radix, 10, u % 256, 0, 0, 0, n, 0, 0, 0, t])
+    for i in range(10):
+        Q = tweak + b'\0'*((-t-b-1) % 16) + bytes([i]) + bebs2osp(B)
+        y = beos2ip(aes_cbcmac(key, P + Q)[:d])
+        c = (NUM_2(A)+y) % (1<<u)
+        C = STR_2(u, c)
+        A = B
+        B = C
+    return A + B
+
+# This is not used except by tests.
+def ff1_aes256_decrypt(key, tweak, x):
+    n = len(x)
+    t = len(tweak)
+    assert minlen <= n and n <= maxlen
+    assert t <= maxTlen
+
+    u = n//2; v = n-u
+    assert u == v
+    A = x[:u]; B = x[u:]
+    assert radix == 2
+    b = cldiv(v, 8)
+    d = 4*cldiv(b, 4) + 4
+    assert d <= 16
+    P = bytes([1, 2, 1, 0, 0, radix, 10, u % 256, 0, 0, 0, n, 0, 0, 0, t])
+    for i in range(9, -1, -1):
+        Q = tweak + b'\0'*((-t-b-1) % 16) + bytes([i]) + bebs2osp(A)
+        y = beos2ip(aes_cbcmac(key, P + Q)[:d])
+        c = (NUM_2(B)-y) % (1<<u)
+        C = STR_2(u, c)
+        B = A
+        A = C
+    return A + B
+
+def test_ff1():
```

Comment:
Add the two test cases from my Rust implementation in here: https://github.com/str4d/fpe/blob/master/src/ff1.rs#L720-L758